### ISSUE TITLE
GaussianSplatting densification desyncs ParameterCount and crashes training on the next step

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#2089]
## Summary

`GaussianSplatting` densification mutates the Gaussian cloud without invalidating the cached parameter layout, so `ParameterCount` keeps reporting the pre-densification size. The next training step publishes a gradient sized from the live cloud, the two disagree, and training dies:

```
System.ArgumentException : Flat gradient length (224) must match ParameterCount (112). (Parameter 'gradients')
   at AiDotNet.NeuralNetworks.NeuralNetworkBase`1.PublishFlatParameterGradients(Vector`1 gradients)   NeuralNetworkBase.cs:4922
   at AiDotNet.NeuralRadianceFields.Models.GaussianSplatting`1.ApplyImageGradients(...)               GaussianSplatting.cs:3031
   at AiDotNet.NeuralRadianceFields.Models.GaussianSplatting`1.Train(Tensor`1, Tensor`1)              GaussianSplatting.cs:2107
```

**`EnableDensification` defaults to `true`** (`GaussianSplattingOptions.cs:37`), so this is reachable with default options — densification is the mechanism the 3DGS paper depends on, and turning it on is the normal case.

## The arithmetic confirms the diagnosis

At `ShDegree = 0` each Gaussian contributes 14 parameters — the generated `GetExtraTrainableTensors` yields `Position(3) + Rotation(4) + Scale(3) + OpacityStorage(1) + Color(3)`.

| Value | Params | Gaussians |
|---|---:|---:|
| Gradient length built from the live `_gaussians` list | 224 | **16** |
| `ParameterCount` reported in the same call | 112 | **8** |

The split had doubled the cloud; the count still described the cloud from before. Both numbers are read inside one `ApplyImageGradients` call, so this is not a timing artifact.

## Root cause

`DensifyAndPrune` ends with:

```csharp
MarkSpatialIndexDirty();
```

That invalidates the renderer's spatial index and nothing else. It never calls `InvalidateParameterCountCache()`, which is what nulls `_parameterLayoutCache` — and `ParameterCount` is derived from `ParameterLayout` (`NeuralNetworkBase.cs:2609`).

This model's parameters *are* the Gaussians, so adding or removing one resizes the parameter surface exactly the way adding or removing a layer does. `InvalidateParameterCountCache`'s own doc says: *"Call this method whenever layers are added, removed, or modified."*

Worth noting this is the **same class of defect** the note at `NeuralNetworkBase.cs:501-508` describes — the `_cachedParameterCount` field removed in #1832 was removed because it "never invalidated when a Layer's ParameterCount grew mid-training via lazy input-shape resolution — leaving ParameterCount reporting a stale pre-resolution size while GetParameters returned the resolved (larger) post-training size." Same failure, reached through the extra-tensors registry instead of layers.

## Why no test caught it

Every existing test avoids the one combination that breaks:

- `GaussianSplattingDensificationTests` drives the internal `RunDensifyAndPruneForTest` hook, proving the split/prune maths and the schedule-window gates — but never that the training loop reaches them.
- `ImageTrainingPathTests.GaussianSplatting_TrainOnImageBatch_ReturnsFiniteLoss` explicitly sets `EnableDensification = false`.

So densification could be wired, configurable, serialized and unit-tested while never surviving a training run, and the suite stayed green.

## Fix

One line at the end of `DensifyAndPrune`:

```csharp
MarkSpatialIndexDirty();
InvalidateParameterCountCache();
```

Plus a regression test that trains through the real image-space path with densification enabled and asserts the cloud changed size, paired with a control arm (densification disabled) asserting it did not — so the assertion cannot pass on ordinary training churn.

Verified: the new tests pass, and the NeuralRadianceFields / GaussianSplatting suites are 110 passed / 0 failed / 1 pre-existing skip.

Fix and tests ship in the consolidated v1 audit PR.

## Repro

```csharp
var model = new GaussianSplatting<float>(new GaussianSplattingOptions {
    ShDegree = 0, EnableSpatialIndex = false, EnableDensification = true, MaxGaussians = 64,
    DensificationStartIteration = 1, DensificationEndIteration = 10_000, DensificationInterval = 1,
    OpacityPruneThreshold = 0.005, GradientNormThreshold = 1e-9,
});
var loader = ImageTrainingDataLoaders.FromViews(views, seed: 17);
for (int i = 0; i < 8; i++)
    model.TrainOnImageBatch(loader, raysPerBatch: 4, optimizerOptions: null);  // throws once a split fires
```

---

*Found during the full re-audit of all 45 open issues against master on 2026-09-05, while verifying #1835. Triaged **v1 Tier 1**.*
